client/rpc-spec-v2/src/chain_head/subscription.rs

```diff
@@ -20,266 +20,1045 @@
+
+		let to_remove: Vec<_> = self
+			.subs
+			.iter_mut()
+			.filter_map(|(sub_id, sub)| {
+				let sub_time = sub.oldest_block_timestamp();
+				// Subscriptions older than the specified pin duration should be removed.
+				let should_remove = match now.checked_duration_since(sub_time) {
+					Some(duration) => duration > self.local_max_pin_duration,
+					None => true,
+				};
+				should_remove.then(|| sub_id.clone())
+			})
+			.collect();
+
+		let mut is_terminated = false;
+		for sub_id in to_remove {
+			if sub_id == request_sub_id {
+				is_terminated = true;
+			}
+			self.remove_subscription(&sub_id);
+		}
+
+		// Make sure we have enough space after first pass of terminating subscriptions.
+		if self.global_blocks.len() < self.global_max_pinned_blocks {
+			return is_terminated
+		}
+
+		// Sanity check: cannot uphold `chainHead` guarantees anymore. We have not
+		// found any subscriptions that have older pinned blocks to terminate.
+		let to_remove: Vec<_> = self.subs.keys().map(|sub_id| sub_id.clone()).collect();
+		for sub_id in to_remove {
+			if sub_id == request_sub_id {
+				is_terminated = true;
+			}
+			self.remove_subscription(&sub_id);
+		}
+		return is_terminated
+	}
+
+	fn pin_block(
+		&mut self,
+		sub_id: &str,
+		hash: Block::Hash,
+	) -> Result<bool, SubscriptionManagementError> {
+		let Some(sub) = self.subs.get_mut(sub_id) else {
+			return Err(SubscriptionManagementError::SubscriptionAbsent)
+		};
+
+		// Block was already registered for this subscription and therefore
+		// globally tracked.
+		if !sub.register_block(hash) {
+			return Ok(false)
+		}
+
+		// Ensure we have enough space only if the hash is not globally registered.
+		if !self.global_blocks.contains_key(&hash) {
+			// Subscription ID was terminated while ensuring enough space.
+			if self.ensure_block_space(sub_id) {
 				return Err(SubscriptionManagementError::ExceededLimits)
 			}
 		}
 
-		Ok(inner.blocks.insert(hash))
+		self.global_pin_block(hash)?;
+		Ok(true)
 	}
 
-	/// Unpin a new block for the current subscription ID.
-	///
-	/// Returns whether the value was present in the set.
-	pub fn unpin_block(&self, hash: &Block::Hash) -> bool {
-		let mut inner = self.inner.write();
-		inner.blocks.remove(hash)
+	fn global_pin_block(&mut self, hash: Block::Hash) -> Result<(), SubscriptionManagementError> {
+		match self.global_blocks.entry(hash) {
+			Entry::Occupied(mut occupied) => {
+				*occupied.get_mut() += 1;
+			},
+			Entry::Vacant(vacant) => {
+				self.backend
+					.pin_block(hash)
+					.map_err(|err| SubscriptionManagementError::Custom(err.to_string()))?;
+
+				vacant.insert(1);
+			},
+		};
+		Ok(())
 	}
 
-	/// Check if the block hash is present for the provided subscription ID.
-	///
-	/// Returns `true` if the set contains the block.
-	pub fn contains_block(&self, hash: &Block::Hash) -> bool {
-		let inner = self.inner.read();
-		inner.blocks.contains(hash)
+	fn global_unpin_block(&mut self, hash: Block::Hash) {
```

Comment:
Something I find a little hard to follow is that  none of these fns are `pub` etc for a bunch of these things, and so it's hard to know which are used by the "parent" interface and which are private and only called from other fns on that type.

I wonder whether moving some of this stuff into separate files would help make the inner and outer interfaces super obvious so that it's easier to see what can and can't be called by the other types?